File: common/test/java/com/googlecode/webdriver/DragAndDropTest.java
package com.googlecode.webdriver;

import java.awt.Point;

public class DragAndDropTest extends AbstractDriverTestCase {
    
    public void testDragAndDrop() throws Exception {
        driver.get(dragAndDropPage);
        RenderedWebElement img = (RenderedWebElement) driver.findElement(By.id("test1"));
        Point expectedLocation = img.getLocation();
        drag(img, expectedLocation, 500, 300);
        assertEquals(expectedLocation, img.getLocation());
        driver.manage().setMouseSpeed(Speed.SLOW);
        drag(img, expectedLocation, -100, -50);
        assertEquals(expectedLocation, img.getLocation());
        driver.manage().setMouseSpeed(Speed.MEDIUM);
        drag(img, expectedLocation, 0, 0);
        assertEquals(expectedLocation, img.getLocation());
        driver.manage().setMouseSpeed(Speed.FAST);
        drag(img, expectedLocation, 1, -1);
        assertEquals(expectedLocation, img.getLocation());
    }
    
    public void testDragAndDropToElement() {
        driver.get(dragAndDropPage);
        RenderedWebElement img1 = (RenderedWebElement) driver.findElement(By.id("test1"));
        RenderedWebElement img2 = (RenderedWebElement) driver.findElement(By.id("test2"));
        img1.dragAndDropBy(100, 100);
        img2.dragAndDropOn(img1);
        assertEquals(img1.getLocation(), img2.getLocation());
    }
    
    public void testElementInDiv() {
        driver.get(dragAndDropPage);
        RenderedWebElement img = (RenderedWebElement) driver.findElement(By.id("test3"));
        Point expectedLocation = img.getLocation();
        drag(img, expectedLocation, 100, 100);
        assertEquals(expectedLocation, img.getLocation());
    }
    
    public void testDragTooFar() {
        driver.get(dragAndDropPage);
        RenderedWebElement img = (RenderedWebElement) driver.findElement(By.id("test1"));
        Point expectedLocation = img.getLocation();
        
        img.dragAndDropBy(Integer.MIN_VALUE, Integer.MIN_VALUE);
        assertEquals (new Point(0, 0), img.getLocation());
        
        img.dragAndDropBy(Integer.MAX_VALUE, Integer.MAX_VALUE);
        //We don't know where the img is dragged to , but we know it's not too
        //far, otherwise this function will not return for a long long time
    }
    
    public void testMouseSpeed() throws Exception {
        driver.get(dragAndDropPage);
        driver.manage().setMouseSpeed(Speed.SLOW);
        assertEquals(Speed.SLOW, driver.manage().getMouseSpeed());
        driver.manage().setMouseSpeed(Speed.MEDIUM);
        assertEquals(Speed.MEDIUM, driver.manage().getMouseSpeed());
        driver.manage().setMouseSpeed(Speed.FAST);
        assertEquals(Speed.FAST, driver.manage().getMouseSpeed());
    }
    
    private void drag(RenderedWebElement elem, Point expectedLocation, 
            int moveRightBy, int moveDownBy) {
        elem.dragAndDropBy(moveRightBy, moveDownBy);
        expectedLocation.move(expectedLocation.x + moveRightBy, expectedLocation.y + moveDownBy);
    }
}

File: common/test/csharp/webdriver-common-test/DragAndDropTest.cs
using System;
using System.Collections.Generic;
using System.Text;
using NUnit.Framework;
using System.Drawing;

namespace OpenQA.Selenium
{
    [TestFixture]
    public class DragAndDropTest : DriverTestFixture
    {
        [Test]
        [Category("Javascript")]
        [IgnoreBrowser(Browser.HtmlUnit)]
        [IgnoreBrowser(Browser.Chrome)]
        public void DragAndDrop()
        {
            driver.Url = dragAndDropPage;
            IRenderedWebElement img = (IRenderedWebElement)driver.FindElement(By.Id("test1"));
            Point expectedLocation = drag(img, img.Location, 150, 200);
            Assert.AreEqual(expectedLocation, img.Location);
            driver.Manage().Speed = Speed.Slow;
            expectedLocation = drag(img, img.Location, -50, -25);
            Assert.AreEqual(expectedLocation, img.Location);
            driver.Manage().Speed = Speed.Medium;
            expectedLocation = drag(img, img.Location, 0, 0);
            Assert.AreEqual(expectedLocation, img.Location);
            driver.Manage().Speed = Speed.Fast;
            expectedLocation = drag(img, img.Location, 1, -1);
            Assert.AreEqual(expectedLocation, img.Location);
        }

        [Test]
        [Category("Javascript")]
        [IgnoreBrowser(Browser.HtmlUnit)]
        [IgnoreBrowser(Browser.Chrome)]
        public void DragAndDropToElement()
        {
            driver.Url = dragAndDropPage;
            IRenderedWebElement img1 = (IRenderedWebElement)driver.FindElement(By.Id("test1"));
            IRenderedWebElement img2 = (IRenderedWebElement)driver.FindElement(By.Id("test2"));
            img2.DragAndDropOn(img1);
            Assert.AreEqual(img1.Location, img2.Location);
        }

        [Test]
        [Category("Javascript")]
        [IgnoreBrowser(Browser.HtmlUnit)]
        [IgnoreBrowser(Browser.Chrome)]
        public void ElementInDiv()
        {
            driver.Url = dragAndDropPage;
            IRenderedWebElement img = (IRenderedWebElement)driver.FindElement(By.Id("test3"));
            Point expectedLocation = drag(img, img.Location, 100, 100);
            Assert.AreEqual(expectedLocation, img.Location);
        }

        [Test]
        [Category("Javascript")]
        [IgnoreBrowser(Browser.IE, "Dragging too far in IE causes the element not to move, instead of moving to 0,0.")]
        [IgnoreBrowser(Browser.HtmlUnit)]
        [IgnoreBrowser(Browser.Chrome)]
        public void DragTooFar()
        {
            driver.Url = dragAndDropPage;
            IRenderedWebElement img = (IRenderedWebElement)driver.FindElement(By.Id("test1"));

            // Dragging too far left and up does not move the element. It will be at 
            // its original location after the drag.
            Point originalLocation = new Point(0, 0);
            img.DragAndDropBy(int.MinValue, int.MinValue);
            Assert.AreEqual(originalLocation, img.Location);

            img.DragAndDropBy(int.MaxValue, int.MaxValue);
            //We don't know where the img is dragged to , but we know it's not too
            //far, otherwise this function will not return for a long long time
        }

        [Test]
        [Category("Javascript")]
        [IgnoreBrowser(Browser.HtmlUnit)]
        [IgnoreBrowser(Browser.Chrome)]
        public void MouseSpeed()
        {
            driver.Url = dragAndDropPage;
            driver.Manage().Speed = Speed.Slow;
            Assert.AreEqual(Speed.Slow, driver.Manage().Speed);
            driver.Manage().Speed = Speed.Medium;
            Assert.AreEqual(Speed.Medium, driver.Manage().Speed);
            driver.Manage().Speed = Speed.Fast;
            Assert.AreEqual(Speed.Fast, driver.Manage().Speed);
        }

        [Test]
        [Category("Javascript")]
        [IgnoreBrowser(Browser.HtmlUnit)]
        [IgnoreBrowser(Browser.IE, "Location property not taking scroll position into account")]
        [IgnoreBrowser(Browser.Chrome)]
        public void ShouldAllowUsersToDragAndDropToElementsOffTheCurrentViewPort()
        {
            driver.Url = dragAndDropPage;

            IJavaScriptExecutor js = (IJavaScriptExecutor)driver;
            int height = Convert.ToInt32(js.ExecuteScript("return window.outerHeight;"));
            int width = Convert.ToInt32(js.ExecuteScript("return window.outerWidth;"));
            js.ExecuteScript("window.resizeTo(300, 300);");

            try
            {
                driver.Url = dragAndDropPage;
                IRenderedWebElement img = (IRenderedWebElement)driver.FindElement(By.Id("test3"));
                Point expectedLocation = drag(img, img.Location, 100, 100);
                Assert.AreEqual(expectedLocation, img.Location);
            }
            finally
            {
                js.ExecuteScript("window.resizeTo(arguments[0], arguments[1]);", width, height);
            }
        }

        private Point drag(IRenderedWebElement elem, Point initialLocation, int moveRightBy, int moveDownBy)
        {
            Point expectedLocation = new Point(initialLocation.X, initialLocation.Y);
            elem.DragAndDropBy(moveRightBy, moveDownBy);
            expectedLocation.Offset(moveRightBy, moveDownBy);
            return expectedLocation;
        }
    }
}
